synapse/storage/databases/main/events.py

```diff
@@ -487,6 +487,14 @@ def _persist_events_txn(
         # NB: This function invalidates all state related caches
         self._update_current_state_txn(txn, state_delta_for_room, min_stream_order)
 
+        # Flag the stream change cache for this room if we added non-backfilled events
+        if max_stream_order > 0:
+            for event, _ in events_and_contexts:
+                assert event.internal_metadata.stream_ordering is not None
+                self.store._events_stream_cache.entity_has_changed(
+                    event.room_id, event.internal_metadata.stream_ordering
+                )
```

Comment:
Hm we do indeed. Could it be the async event cache changes? Specifically has [this addition](https://github.com/matrix-org/synapse/commit/2ee0b6ef4b78bada535beb30301cf0e01cbb7d81#diff-f823e25aa66295818eacd76862f7827607893462dcfb7854448578de6ae365e2R860-R862) opened a gap here because the await call will mean the invalidation isn't handled synchronously?

That might explain the single-monolith synapse issue seen in testing but not where the tokens are out of sync between workers...